src/renderer/sw_engine/tvgSwStroke.cpp

```diff
@@ -374,8 +374,8 @@ static void _lineTo(SwStroke& stroke, const SwPoint& to)
 {
     auto delta = to - stroke.center;
 
-    //a zero-length lineto is a no-op; avoid creating a spurious corner
-    if (delta.zero()) return;
+    //a zero-length lineto is a no-op if stroke width is 0; avoid creating a spurious corner
+    if (delta.zero() && stroke.width == 0) return;
```

Comment:
In this place (also in l.454), ```stroke.width == 0``` is never true because it was already checked [here](https://github.com/thorvg/thorvg/blob/285d1427ad823e91c7f3915629a50c4376d84759/src/renderer/sw_engine/tvgSwRenderer.cpp#L96)

What do you think about adding the following check in the ```strokeParseOutline()``` function:
```if (!nonZeroLen && stroke->cap != StrokeCap::Butt)```  
for each contour, i.e., each segment starting with M - (after 'close:' in l.879)
If such a condition is met, it would mean that the cap should be drawn. 
I did some tests by calling the ```_firstSubPath()``` function if the above is true, but it has to be double checked

Reply:
```nonZeroLen``` is the logical sum of the booleans returned from _lineTo and _cubicTo :
```
nonZeroLen |= _lineTo(*stroke, *pt);
nonZeroLen |= _cubicTo(...)
```

Reply:
Could you please suggest how exactly to achieve this? Unfortunately, I’m not familiar enough with the project to accomplish it without issues.

The proposed _lineTo function in tvgSwStroke.cpp does not return a bool, while its variant that does is located in tvgSwRle.cpp. However, it requires a parameter of type RleWorker, which, unfortunately, I’m not familiar with.

Reply:
If I try this kind of hacky solution to determine nonZeroLength:

```c++
close:  
    auto nonZeroLength = start != stroke->center;  
    if (!nonZeroLength && stroke->cap != StrokeCap::Butt)  
        _firstSubPath(*stroke, stroke->angleIn, 0);  
    if (!stroke->firstPt)  
        _endSubPath(*stroke);  
    first = last + 1;  
}
```

Some minimal examples display correctly, while others do not.

![image](https://github.com/user-attachments/assets/e6ea98af-84b2-4e49-bbf6-128508fbaf5d)
![image](https://github.com/user-attachments/assets/9ad9c43e-a209-4ea0-b95d-4cb2002272e2)

Reply:
Can you try this?:
```
        auto drawn = false;
        while (pt < limit) {
            //emit a single line_to
            if (types[0] == SW_CURVE_TYPE_POINT) {
                ++pt;
                ++types;
                drawn |= _lineTo(*stroke, *pt);
            //types cubic
            } else {
                pt += 3;
                types += 3;
                if (pt <= limit) drawn |= _cubicTo(*stroke, pt[-2], pt[-1], pt[0]);
                else if (pt - 1 == limit) drawn |= _cubicTo(*stroke, pt[-2], pt[-1], start);
                else goto close;
            }
        }
    close:
        if (!drawn && stroke->cap != StrokeCap::Butt) _firstSubPath(*stroke, 0, 0);
        if (!stroke->firstPt) _endSubPath(*stroke);
        first = last + 1;
```
this is part of the ```strokeParseOutline``` fun. 

you also have to change _lineTo and _cubicTo (from the tvgSwStroke.cpp) to return bool - in the same places you edited before:

in _cubicTo:
```
        //ignoreable size
        if (valid < 0 && arc == bezStack) {
            stroke.center = to;
            return false;
        }
```
otherwise return true

and in _lineTo:
```
    if (delta.zero()) return false;
```
otherwise return true

please let me know if this doesn't help and send a sample if possible that is problematic (dashed and trimmed stroke is a different story, I can solve it after we handle this issue)

Reply:
WOW. After these changes, it seems that the problem is solved. Thank you so much for the support.
Can we thoroughly test the implemented changes and merge them into the main branch of the project?